internal/rbd/rbd_healer.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
```

Comment:
In my opinion this is more similar to the `internal/controller` package. The `rbd` package talks directly to the Ceph cluster and maintains RBD images there. The healer inspects Kubernetes objects (VolumeAttachment) and calls a CSI function (NodeStageVolume). The current `rbd` package is already very large, and ideally should get split into functions for `rbdImage` and separate CSI servers that call a clean `rbdImage`-API.

Having the healer as much as self-contained as possible, will make it easier to move it to csi-addons later.

Reply:
I wonder moving healer into a different package will lead to cross importing b/w rbd and healer packages.

Because healer needs access to NodeServer structure and NodeStageVolume() functions from rbd, the same way rbd package driver.go needs access to runVolumeHealer() function from healer package.

Does that explain the dependency?

Reply:
hmm, maybe it is indeed problematic due to the immense tightness of the NodeServer+rbdImage. We really should split out the rbdImage functions and CSI servers into their own packages.